### PR TITLE
Automatic doc generation for major binaries

### LINKED ISSUE #8493: Auto-generate user manuals for all kube binaries, not just kubectl
When reading user manual for kubelet. I found it contains very few options than kubelet provides, and also some errors. For example, '--etcd_servers' is no longer supported. 

Same problem for manuals of other kube binaries except kubectl.

I'm trying to fix this and re-generate the man files. I noticed there're some options used for testing. Shall I keep them in the md files or remove them?

### PR DESCRIPTION
**Pull request by eparis**
Fixes #8493.

Converts apiserver, controller-manager, scheduler, proxy, kubelet to spf13/cobra. It then uses cobra to generate md, man, and bash completions. We currently do this for kubectl, this does it for more.

The one thing to note is that cobra `Run()` functions do not return an error. Thus I started using `glog.Fatal()`. This can be undone if need be...

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @mikedanese

**Comment by mikedanese** [comment]
Thanks @eparis! Feel free to assign to me when this is ready.

**Comment by eparis** [comment]
It is ready, but I'd like someone like @smarterclayton or @pweil- to tell me if removing the error return to `Run()` is going to be a big pain for them?

**Comment by eparis** [comment]
https://github.com/kubernetes/kubernetes/pull/13018/files#diff-36e1d7088209b1ab6ff878b18bb35210R298

**Comment by smarterclayton** [comment]
Run is a big problem for us.

**Comment by smarterclayton** [comment]
We need all the errors for most of the startup stuff - glog.Fatalf really needs to only be in main packages and the cobra stub

**Comment by eparis** [comment]
Ok, I made no changes to `Run()` or any returns. I should point out, every single one uses `glog.Fatalf()` in some places. Someone likely will want to look into those if OpenShift doesn't like them...

**Comment by smarterclayton** [comment]
It's not really about Openshift and code reuse as much as it is that using glog.Fatalf means that code can't be tested in a unit test.  We removed the vast majority of them from the cli, but did not clean this path.  I'll open an issue to track the rest and make reviewers aware.

**Comment by bgrant0607** [comment]
Is there a separate issue for returning errors from Run?

Is this still worth doing given that we're trying to eliminate all these flags and switch to configuration files?

**Comment by eparis** [comment]
@bgrant0607 are you asking if this is worthwhile at all?  It's very little code change for consistent docs. Even if we don't want to autogenerate for kube-*, i'll likely still want to push about 80% of the code changes for cleaner generation for kubectl, which isn't going to go away...

**Comment by bgrant0607** [comment]
Ok

**Comment by mikedanese** [comment]
Thanks, I'll review tomorrow or this weekend.

**Comment by eparis** [comment]
ok to test

**Comment by eparis** [comment]
ok to test

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -68,7 +72,7 @@ type ProxyServer struct {
 
 // NewProxyServer creates a new ProxyServer object with default parameters
 func NewProxyServer() *ProxyServer {
-	return &ProxyServer{
+	s := ProxyServer{
```
Comment (mikedanese) [inline cmd/kube-proxy/app/server.go]: was this change on purpose?

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -76,6 +80,38 @@ func NewProxyServer() *ProxyServer {
...
+			if err := s.Run(args); err != nil {
+				glog.Fatalf("%v", err)
+			}
+		},
+	}
+	genutils.AddGeneratedDocsCommands(cmd)
+
+	s.AddFlags(cmd.Flags())
+	pflag.CommandLine.AddGoFlagSet(goflag.CommandLine)
+	cmd.SetGlobalNormalizationFunc(util.WordSepNormalizeFunc)
```
Comment (mikedanese) [inline cmd/kube-proxy/app/server.go]: Is InitFlags unused now? Can we refactor it to do what's done here and repeated in other flags? If not, can we remove it?

File: plugin/cmd/kube-scheduler/scheduler.go [hunk 3]
```diff
@@ -33,14 +30,11 @@ func init() {
...
-	util.InitFlags()
+	cmd := app.NewSchedulerServerCommand()
+
 	util.InitLogs()
 	defer util.FlushLogs()
 
-	verflag.PrintAndExitIfRequested()
-
-	s.Run(pflag.CommandLine.Args())
+	cmd.Execute()
```
Comment (mikedanese) [inline plugin/cmd/kube-scheduler/scheduler.go]: I'm not sure what would cause a cobra command to return an error without calling s.Run but glog.Fatal(cmd.Execute()) to be safe?

File: hack/verify-generated-docs.sh [hunk 4]
```diff
@@ -22,8 +22,17 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
 source "${KUBE_ROOT}/hack/lib/init.sh"
 
 kube::golang::setup_env
+binaries=(
```
Comment (mikedanese) [inline hack/verify-generated-docs.sh]: extreme nit: add a space before this line to make the diff match the diff of hack/update-generated-docs.sh

**Comment by mikedanese** [comment]
Small nits but this looks. (this is PR 90% autogenerated). Sorry for slow review. Thanks!

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (eparis) [inline cmd/kube-proxy/app/server.go]: not really on purpose. I originally killed this func, clayton complained/said they needed it, so I restored it. But I restored it by hand, following the same convention as the other New*Server() functions. So it isn't needed, but does make the code look slightly more like the other similar code...

File: cmd/kube-proxy/app/server.go (hunk 2 above)
Comment (eparis) [inline cmd/kube-proxy/app/server.go]: Still used by contrib/mesos and integration. maybe I can keep working on those. Mesos really needs to make the switch to get sanity on their flags...  But separate PR

**Comment by eparis** [comment]
travis is 'realish'   docs are not deterministic. crap.

**Comment by mikedanese** [comment]
LGTM once this is consistently passing tests

**Comment by eparis** [comment]
yeah, not sure what to do. the value of GOMAXPROCS during `go build` leaks into the default value of test.parallel because of https://github.com/kubernetes/kubernetes/issues/13565  Not totally sure how to handle that...

**Comment by eparis** [comment]
I might break this into multiple PRs and get as much in as I can.

**Comment by mikedanese** [comment]
@eparis can this be closed as it's been broken up?

**Comment by mikedanese** [comment]
PR size appears to be nondeterministic. @lavalamp

**Comment by eparis** [comment]
@lavalamp I'll take a look, not sure how this can happen, unless you have stuff in the logs about `.generated_docs`....

**Comment by lavalamp** [comment]
I bet https://github.com/kubernetes/contrib/pull/100 was the cause of it switching to XL a day ago-- let's see if it does it again.

**Comment by mikedanese** [comment]
I think this can be closed.
